framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -267,97 +268,176 @@ protected boolean areEqualInHierarchy(
+     * @return true if inside is contained by outside, or if canBeCovariant == true and {@code
+     *     inside <: outside}
+     */
+    protected boolean isContainedByBoundType(
+            AnnotatedTypeMirror inside, BoundType outside, boolean canBeCovariant) {
+        if (BoundType.isBoundType(inside)) {
+            BoundType insideBoundType = new BoundType(inside);
+            if (!insideBoundType.hasSuperBound && !outside.hasSuperBound) {
+                return (canBeCovariant || isSubtype(outside.lower, insideBoundType.lower))
+                        && isContainedByBoundType(insideBoundType.upper, outside, canBeCovariant);
+            } else if (outside.hasSuperBound
+                    || TypesUtils.isObject(outside.upper.getUnderlyingType())) {
+                return (canBeCovariant || isSubtype(insideBoundType.upper, outside.upper))
+                        && isContainedByBoundType(insideBoundType.lower, outside, canBeCovariant);
             }
-            outsideUpperBound = ((AnnotatedWildcardType) outsideUpperBound).getExtendsBound();
         }
 
-        AnnotatedTypeMirror castedInside =
-                AnnotatedTypes.castedAsSuper(inside.atypeFactory, inside, outsideUpperBound);
-        if (!isSubtypeCaching(castedInside, outsideUpperBound)) {
-            return false;
+        if (BoundType.isBoundType(outside.upper)) {
+            BoundType outsideUpper = new BoundType(outside.upper);
+            Set<AnnotationMirror> setA =
+                    AnnotatedTypes.findEffectiveLowerBoundAnnotations(
+                            qualifierHierarchy, outsideUpper.lower);
+            Set<AnnotationMirror> setB =
+                    AnnotatedTypes.findEffectiveLowerBoundAnnotations(
+                            qualifierHierarchy, outside.lower);
+            Set<? extends AnnotationMirror> glb =
+                    qualifierHierarchy.greatestLowerBounds(setA, setB);
+            addToLowestBound(outside.lower, glb);
+            return isContainedByBoundType(inside, outsideUpper, canBeCovariant);
+        } else if (BoundType.isBoundType(outside.lower)) {
+            BoundType outsideLower = new BoundType(outside.lower);
+            return isContainedByBoundType(inside, outsideLower, canBeCovariant);
+        } else {
+            if (canBeCovariant) {
+                if (outside.hasSuperBound) {
+                    return isSubtype(outside.lower, inside);
+                } else {
+                    return isSubtype(inside, outside.upper);
+                }
+            }
+            return isSubtype(outside.lower, inside) && isSubtype(inside, outside.upper);
         }
+    }
 
-        if (outside.getKind() == TypeKind.WILDCARD
-                && outsideLowerBound.getKind() == TypeKind.TYPEVAR) {
-            // tests/all-systems/Issue1991.java crashes without this.
-            return true;
+    /**
+     * Add {@code glb} to the lowest bound of {@code type}.
```

Comment:
I cannot follow this comment.  Is `glb` the same as `annos`?  Please define "lowest bound".  (I see what it is from the code, but clarifying the intent will help to ensure that the two are consistent.)